### ISSUE TITLE
Updating network weights using only a specific portion of batch samples

### ISSUE THREAD

**Issue by AminHP** [keras-team/keras#17493]
**System information**.

TensorFlow version (you are using):  2.11.0
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

In the `train_step` method of the `Model` class, all samples in the batch are used for updating the network weights. In many problems, this is the correct and common way to update weights. However, in some situations, there are other ways for updating network weights. Sometimes, we can compute loss using all samples in the batch but update the network weights using only a subset of that batch. This is useful when we need to have a large batch but the memory limits don't allow us to compute the gradients for all samples. Hence, a solution is to compute the loss value with all batch samples, and then compute the gradients and update network weights via a subset of those samples. In most problems, this is not meaningful, but in Metric Learning and similar problems, it is quite useful.

For example, consider we are using the `MultiSimilarityLoss` in [this example](https://keras.io/examples/vision/metric_learning_tf_similarity). In this loss function, each batch contains **K** samples from **C** classes. The purpose of `MultiSimilarityLoss` consists in computing loss in a way that the **K** samples belonging to a class get close in the feature space, while at the same time these **K** samples get distant from other classes' samples in that batch. This loss function needs a large batch to perform satisfactorily. However, computing the gradients for all samples in a large batch wouldn't fit in memory. A simple solution, as explained, is computing gradients and updating network weights for only **m** classes from those **C** classes in the batch. Although the loss value is calculated based on all samples in the batch, it is possible to update weights based on a portion of that batch.

**Will this change the current api? How?**

Probably adds a parameter, named `batch_compute_gradient_portion`, to the `Model.fit` or `Model.__init__`.

**Who will benefit from this feature?**

Problems that need a large batch in order to reduce loss effectively, when it is not possible to compute gradients for all samples in a large batch due to GPU memory limits.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes

- Briefly describe your candidate solution(if contributing):

A possible solution is implemented below. Note that I've tested it in an enterprise project for a company I work for, and it's been successful.

```python
class Model(tf.keras.models.Model):

    def __init__(self, *args, **kwargs):
        self.batch_compute_gradient_portion = float(kwargs.pop('batch_compute_gradient_portion', 1))
        assert 0. < self.batch_compute_gradient_portion <= 1.

        super().__init__(*args, **kwargs)

    def train_step(self, data):
        x, y, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)

        l = tf.cast(tf.shape(x)[0], tf.float32)
        k = tf.cast(self.batch_compute_gradient_portion * l, tf.int32)

        # Run forward pass.
        y_pred_without_gradient = self(x[k:], training=True)

        with tf.GradientTape() as tape:
            y_pred_with_gradient = self(x[:k], training=True)

            y_pred = tf.concat([y_pred_with_gradient, y_pred_without_gradient], axis=0)

            loss = self.compute_loss(x, y, y_pred, sample_weight)

        self._validate_target_and_loss(y, loss)

        # Run backwards pass.
        self.optimizer.minimize(loss, self.trainable_variables, tape=tape)

        return self.compute_metrics(x, y, y_pred, sample_weight)
```

**Comment by rchao** [comment]
Question: is it possible to use a smaller batch size to avoid memory issue? We're unsure if this is generalized enough to be made in the built-in training loop, so would like additional information.

**Comment by AminHP** [comment]
@rchao 
There are two concerns with smaller batch sizes:
1. In order to reduce loss in an effective and efficient way in metric learning, the model needs to compare each sample with a large number of samples from the same and other classes at the same time. The situation is similar to solving a classification problem with a batch size of 1, which is really difficult and time-consuming to train and retains a lot of fluctuations in the training process.
2. Training with only a portion of the batch is not just about memory issues. It helps to avoid overfitting. I've tested multiple scenarios in which the batch size was kept constant, but the `batch_compute_gradient_portion` value changed from 0.25 to 1. The models trained with `batch_compute_gradient_portion < 1` were clearly less influenced by overfitting. Their generalization ability in predicting completely new samples with different distributions was significantly higher.

**Comment by AminHP** [comment]
Any updates @rchao ?

**Comment by AminHP** [comment]
Yes that works. But the problem is we have to pass `batch_compute_gradient_portion` to the model's constructor, which is not a clean way. I was thinking it is better to be a parameter in the `fit` function. wdyt @rchao ?

**Comment by AminHP** [comment]
According to my explanations, is this feature generalized enough to be added to the repo @rchao ? It would be much appreciated if you could give the final decision on this feature request.

**Comment by rchao** [comment]
Hello @AminHP , thanks for the information! I tend to be a bit more conservative on this as the impact on the code complexity is apparent and the current training library isn't blocking users from having their custom training step. Hopefully this can work for your use case and thanks for your understanding.

**Comment by AminHP** [comment]
Thanks for your consideration.
